File: rolling_hash.cs
public struct RollingHash
{
    private long _hashA;
    private long _hashB;
    private int _length;

    public long HashA => _hashA;
    public long HashB => _hashB;
    public int Length => _length;

    public static RollingHash Zero => new(0, 0, 0);

    public RollingHash()
    {
        _hashA = 0;
        _hashB = 0;
        _length = 0;
    }

    public RollingHash(long hashA, long hashB, int length)
    {
        _hashA = hashA;
        _hashB = hashB;
        _length = length;
    }

    public static bool operator ==(RollingHash left, RollingHash right)
    {
        return left._hashA == right._hashA && left._hashB == right._hashB && left._length == right._length;
    }

    public static bool operator !=(RollingHash left, RollingHash right)
    {
        return !(left == right);
    }

    public override int GetHashCode()
    {
        return (_hashA, _hashB).GetHashCode();
    }

    public override bool Equals([NotNullWhen(true)] object obj)
    {
        if (obj is RollingHash other)
        {
            return this == other;
        }
        else
        {
            return false;
        }
    }

    public static RollingHash FromChar(char c)
    {
        return new (c, c, 1);
    }
}

public static class RabinKarp
{
    public const long BaseA = 3491;
    public const long ModA = 481840747;
    public const long BaseB = 8761;
    public const long ModB = 999750347;

    private static long[] _powerA;
    private static long[] _powerB;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static long SafeMod(long a, long m)
    {
        return ((a % m) + m) % m;
    }

    public static void Setup(int maxLength)
    {
        _powerA = new long[maxLength + 1];
        _powerB = new long[maxLength + 1];
        _powerA[0] = 1L;
        _powerB[0] = 1L;
        for (int i = 1; i <= maxLength; i++)
        {
            _powerA[i] = (_powerA[i - 1] * BaseA) % ModA;
            _powerB[i] = (_powerB[i - 1] * BaseB) % ModB;
        }
    }

    public static RollingHash PrefixDiff(RollingHash left, RollingHash right)
    {
        int diff = right.Length - left.Length;
        long hashA = SafeMod(right.HashA - (left.HashA * _powerA[diff]) % ModA, ModA);
        long hashB = SafeMod(right.HashB - (left.HashB * _powerB[diff]) % ModB, ModB);

        return new(hashA, hashB, diff);
    }

    public static RollingHash Concat(RollingHash left, RollingHash right)
    {
        int len = left.Length + right.Length;
        long hashA = ((left.HashA * _powerA[right.Length]) % ModA + right.HashA) % ModA;
        long hashB = ((left.HashB * _powerB[right.Length]) % ModB + right.HashB) % ModB;
        return new(hashA, hashB, len);
    }

    public static RollingHash HashFromString(string s)
    {
        long hashA = 0L;
        long hashB = 0L;
        for (int i = 0; i < s.Length; i++)
        {
            hashA = (hashA * BaseA) % ModA;
            hashA += s[i];
            hashA %= ModA;

            hashB = (hashB * BaseB) % ModB;
            hashB += s[i];
            hashB %= ModB;
        }

        return new(hashA, hashB, s.Length);
    }

    public static long PowerA(int exp)
    {
        return _powerA[exp];
    }

    public static long PowerB(int exp)
    {
        return _powerB[exp];
    }
}

public sealed class RollingHashString
{
    private int _length;
    private string _source;
    private RollingHash[] _prefix;

    public int Length => _length;
    public string Source => _source;

    public RollingHashString(string source)
    {
        _source = source;
        _length = source.Length;

        _prefix = new RollingHash[_length + 1];
        _prefix[0] = RollingHash.Zero;
        for (int i = 1; i <= _length; i++)
        {
            _prefix[i] = RabinKarp.Concat(_prefix[i - 1], RollingHash.FromChar(_source[i - 1]));
        }
    }

    public RollingHash GetHash(int l, int r)
    {
        return RabinKarp.PrefixDiff(_prefix[l], _prefix[r]);
    }

    public RollingHash GetPrefixHash(int length)
    {
        return _prefix[length];
    }
}

File: cpp/rolling_hash.cpp
struct RollingHash
{
private:
    ll _hashA;
    ll _hashB;
    int _length;

public:
    RollingHash()
    {
        _hashA = 0;
        _hashB = 0;
        _length = 0;
    }

    RollingHash(ll hashA, ll hashB, int length)
    {
        _hashA = hashA;
        _hashB = hashB;
        _length = length;
    }

    ll HashA()
    {
        return _hashA;
    }

    ll HashB()
    {
        return _hashB;
    }

    int Length()
    {
        return _length;
    }

    bool operator ==(const RollingHash& other)
    {
        return _hashA == other._hashA && _hashB == other._hashB && _length == other._length;
    }

    inline static RollingHash Zero()
    {
        return RollingHash(0, 0, 0);
    }

    inline static RollingHash FromChar(char c)
    {
        return RollingHash((ll)c, (ll)c, 1);
    }
};

class RabinKarp
{
public:
    static const ll BaseA = 3491;
    static const ll ModA = 481840747;
    static const ll BaseB = 8761;
    static const ll ModB = 999750347;

private:
    static vector<ll> _powerA;
    static vector<ll> _powerB;

public:
    static inline ll SafeMod(ll a, ll m)
    {
        return ((a % m) + m) % m;
    }

    static void Setup(int maxLength)
    {
        _powerA.resize(maxLength + 1);
        _powerB.resize(maxLength + 1);
        _powerA[0] = 1LL;
        _powerB[0] = 1LL;
        for (int i = 1; i <= maxLength; i++)
        {
            _powerA[i] = (_powerA[i - 1] * BaseA) % ModA;
            _powerB[i] = (_powerB[i - 1] * BaseB) % ModB;
        }
    }

    static RollingHash PrefixDiff(RollingHash left, RollingHash right)
    {
        int diff = right.Length() - left.Length();
        ll hashA = SafeMod(right.HashA() - (left.HashA() * _powerA[diff]) % ModA, ModA);
        ll hashB = SafeMod(right.HashB() - (left.HashB() * _powerB[diff]) % ModB, ModB);

        return RollingHash(hashA, hashB, diff);
    }

    static RollingHash Concat(RollingHash left, RollingHash right)
    {
        int len = left.Length() + right.Length();
        ll hashA = ((left.HashA() * _powerA[right.Length()]) % ModA + right.HashA()) % ModA;
        ll hashB = ((left.HashB() * _powerB[right.Length()]) % ModB + right.HashB()) % ModB;
        return RollingHash(hashA, hashB, len);
    }

    static RollingHash HashFromString(string s)
    {
        ll hashA = 0LL;
        ll hashB = 0LL;
        for (int i = 0; i < s.size(); i++)
        {
            hashA = (hashA * BaseA) % ModA;
            hashA += (ll)s[i];
            hashA %= ModA;

            hashB = (hashB * BaseB) % ModB;
            hashB += (ll)s[i];
            hashB %= ModB;
        }

        return RollingHash(hashA, hashB, (int)s.size());
    }

    inline static ll PowerA(int exp)
    {
        return _powerA[exp];
    }

    inline static ll PowerB(int exp)
    {
        return _powerB[exp];
    }
};

vector<ll> RabinKarp::_powerA;
vector<ll> RabinKarp::_powerB;

class RollingHashString
{
private:
    int _length;
    string _source;
    vector<RollingHash> _prefix;

public:
    RollingHashString(string source)
    {
        assert(source.size() != 0);
        _source = source;
        _length = (int)source.size();
        // pre-compute prefix hashes

        _prefix.resize(_length + 1);
        _prefix[0] = RollingHash::Zero();
        for (int i = 1; i <= _length; i++)
        {
            _prefix[i] = RabinKarp::Concat(_prefix[i - 1], RollingHash::FromChar(_source[i - 1]));
        }
    }

    RollingHash GetHash(int l, int r)
    {
        assert(0 <= l && l < _length && l <= r && r <= _length);
        return RabinKarp::PrefixDiff(_prefix[l], _prefix[r]);
    }

    RollingHash GetPrefixHash(int length)
    {
        assert(0 <= length && length <= _length);
        return _prefix[length];
    }

    string Source()
    {
        return _source;
    }

    int Length()
    {
        return _length;
    }
};